[18.0][ADD] sale_order_add_byproduct

bosd opened the pull request:
This module is designed for contract manufacturing scenarios where the
client provides raw materials or a base product for processing. Unlike
standard manufacturing where the output quantity of finished goods is
precisely known, this use case involves a process where the exact yield
of **by-products** is uncertain until after the manufacturing operation
is complete.

Scenario
--------

A contract manufacturer processes materials supplied by a client. During
this process, in addition to any primary output, a variable quantity of
by-products is inevitably generated. The client is obligated to purchase
these by-products from the contract manufacturer.

bosd committed:
[ADD] sale_order_add_byproduct

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+        sale_order = self.procurement_group_id.sale_id
+        if not sale_order:
+            # Try to find the sale order by origin
+            if self.origin:
```
Why doing such queries as you have this direct field?

https://github.com/odoo/odoo/blob/18.0/addons/sale_mrp/models/mrp_production.py#L14C5-L14C17

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+            # In Odoo 18, the quantity field might be named differently
+            # Use quantity_done if available, otherwise fallback to quantity
+            # If neither is available, use the move line quantities
+            if (
```
This should be removed as in 18.0, `quantity_done` field does not exist anymore.

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+                and byproduct_move.quantity_done
+            ):
+                quantity_produced = byproduct_move.quantity_done
+            elif hasattr(byproduct_move, "quantity") and byproduct_move.quantity:
```
stock move has `quantity` field, so the `else` case is not relevant anymore.

rousseldenis replied:
In fact, if no move lines, the quantity done == 0.0. But, if you look for a done quantity, you should take the move lines with `picked` field as `True`

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+                    continue
+                quantity_produced = sum(byproduct_move.move_line_ids.mapped("quantity"))
+
+            if not byproduct_move.product_id.sale_ok:
```
This has to be put at the beginning of the loop to avoid too much conditions evaluations.

rousseldenis replied:
Moreover, I would have put this in a separate function like `_get_byproduct_moves_to_add()` or somethng like that.

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+            if not byproduct_move.product_id.sale_ok:
+                continue
+
+            if quantity_produced > 0:
```
Use `float_compare()`

rousseldenis commented on sale_order_add_byproduct/models/res_config_settings.py:
```diff
@@ -0,0 +1,33 @@
+    byproduct_note_template = fields.Char(
+        string="By-product Note Template",
+        help="Template for the note on sale order lines created for by-products. "
+        "You can use placeholders like {product_name} and {mo_name}.",
```
Use `config_parameter="sale_order_add_byproduct.byproduct_note_template"` to avoid `get` and `set` methods

rousseldenis replied:
I'm wondering if a per company parameter shoulde be better or not

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+                    )
+                else:
+                    # Get the template from the config parameter
+                    config_settings = self.env["res.config.settings"].create({})
```
Don't do that. Move the `get_` method (in res.config.settings - see my other comment) to an api.model one here.

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+                        )
+                    else:
+                        note = product_to_add.name
+                    self.env["sale.order.line"].create(
```
You should build create values (a list) in the loop and after the loop, just call one `create()` with that list.

rousseldenis commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,131 @@
+                            "is_mrp_byproduct_line": True,
+                        }
+                    )
+                    _logger.info(
```
Are all these info lines should stay as this will pollute the log file? I would have choose debug instead.

rousseldenis commented on sale_order_add_byproduct/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+        # For regular lines on confirmed orders, ensure procurement is triggered
+        # This ensures the procurement group is properly maintained
+        if record.order_id and record.order_id.state == 'sale':
+            # Only trigger for regular lines, not by-product lines
```
I'm not sure about the code here after as this can lead to unwanted behaviours.

bosd committed:
[IMP] sale_order_add_byproduct: Respect pricelists

bosd committed:
Fixups

bosd committed:
FEATURE: Add company-specific byproduct note template with fallback

- Add res_company model with byproduct_note_template field
- Update res_config_settings to use compute/inverse methods for company-specific settings
- Update mrp_production to use company-specific template with fallback to global config
- Add proper fallback logic: company setting -> global config -> default value
- Maintain backward compatibility with existing global configurations
- Include _get_byproduct_moves_to_add() method for cleaner code structure
- Apply proper float_compare() usage as per review
- Update __init__.py to import new company model

bosd committed:
FIX: Change logger level from info to debug in mrp_production.py

- Change _logger.info() calls to _logger.debug() for internal processing messages
- Maintain appropriate logging level for diagnostic information
- Keep debug level for internal processing logs to reduce noise in production
- Apply to all logging calls in the _add_byproducts_to_sale_order method

bosd commented:
Thanks for the review. 
CI is still red, but that seems to come from a different module.

Dranyel-Bosd reviewed (approved):
LGTM

alexey-pelykh commented on sale_order_add_byproduct/models/sale_order_line.py:
```diff
@@ -0,0 +1,35 @@
+
+        # Call super only on lines that should be procured
+        if lines_not_to_procure:
+            return True
```
**Bug (critical):** This logic is inverted for mixed recordsets. If `self` contains 5 lines and 1 is a byproduct, this returns `True` immediately and **skips procurement for the other 4 lines**.

The correct pattern is to exclude byproduct lines and call `super()` on the remaining lines:

```python
def _action_launch_stock_rule(self, previous_product_uom_qty=False):
    lines_to_procure = self.filtered(lambda line: not line.is_mrp_byproduct_line)
    if not lines_to_procure:
        return True
    return super(SaleOrderLine, lines_to_procure)._action_launch_stock_rule(
        previous_product_uom_qty=previous_product_uom_qty
    )
```

alexey-pelykh commented on sale_order_add_byproduct/models/sale_order_line.py:
```diff
@@ -0,0 +1,35 @@
+        for line in lines:
+            if line.order_id.state == "sale" and not line.is_mrp_byproduct_line:
+                line._action_launch_stock_rule()
+        return lines
```
**Bug:** Overriding `create` to manually call `_action_launch_stock_rule()` for non-byproduct lines is risky. Odoo's standard `sale.order.line` already triggers procurement on line creation for confirmed SOs via its own mechanism. This override may cause **double procurement** (two MOs created for one SO line).

I believe the `_action_launch_stock_rule` override above (once fixed) is sufficient to prevent byproduct lines from triggering procurement. This entire `create` override should likely be removed.

If it IS needed, please add a test that demonstrates the scenario it solves -- currently no test covers the `create`-triggered procurement path specifically.

alexey-pelykh commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,200 @@
+                if not sale_order and self.origin.startswith("S"):
+                    sale_order = self.env["sale.order"].search(
+                        [("name", "=", self.origin)], limit=1
+                    )
```
**Dead code:** This `if` branch (`self.origin.startswith("S")`) executes the exact same search as lines 38-40 above:
```python
sale_order = self.env["sale.order"].search(
    [("name", "=", self.origin)], limit=1
)
```
This duplicate search can never find anything the first one missed. Remove this branch entirely.

alexey-pelykh commented on sale_order_add_byproduct/models/mrp_production.py:
```diff
@@ -0,0 +1,200 @@
+                    if config_param:
+                        note = config_param.format(
+                            product_name=product_to_add.name, mo_name=self.name
+                        )
```
**Security concern:** Using `str.format()` with a user-controlled template string is a known Python injection vector. A malicious admin could craft a template like `{product_name.__class__.__mro__}` to leak internal Python object information.

Consider using `safe_substitute` from `string.Template`, or restrict the format to known placeholders via a simple `.replace()` approach:
```python
note = config_param.replace("{product_name}", product_to_add.name).replace("{mo_name}", self.name)
```

alexey-pelykh commented on sale_order_add_byproduct/models/sale_order_line.py:
```diff
@@ -0,0 +1,35 @@
+from odoo import fields, models
```
**OCA convention:** Missing license header. All Python files in OCA modules must have the copyright and license header:
```python
# Copyright 2025 OBS Solutions BV
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```

alexey-pelykh commented on sale_order_add_byproduct/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+    "license": "AGPL-3",
+    "depends": ["mrp", "sale_management", "sale_stock"],
+    "data": ["data/config_parameter_data.xml", "views/res_config_settings_views.xml"],
+    # "demo": ["demo/demo_data.xml"],
```
**Cleanup:** Remove the commented-out `"demo"` line. OCA guidelines discourage committed commented-out code.

alexey-pelykh commented on sale_order_add_byproduct/models/res_config_settings.py:
```diff
@@ -0,0 +1,29 @@
+
+    byproduct_note_template = fields.Char(
+        string="By-product Note Template",
+        help="Template for the note on sale order lines created for by-products. "
```
As `rousseldenis` already noted: since you now have a `res.company` field, the settings field should use `related="company_id.byproduct_note_template"` rather than manual `_compute` / `_inverse` methods. This is the standard OCA pattern for company-level settings and removes ~15 lines of code.

alexey-pelykh reviewed (changes requested):
## Review: REQUEST_CHANGES

Thank you for the contribution -- the business case is well-documented and the test coverage is solid. However, there are several correctness issues and OCA convention gaps that should be addressed before merge.

### Critical Findings

| # | File | Severity | Finding |
|---|------|----------|---------|
| 1 | `sale_order_line.py` | Bug | `_action_launch_stock_rule` override breaks mixed recordsets -- returns `True` for the entire batch when ANY line is a byproduct, silently skipping procurement for legitimate lines |
| 2 | `sale_order_line.py` | Bug | `create` override may double-trigger procurement for non-byproduct lines added to confirmed SOs (Odoo already handles this) |
| 3 | `mrp_production.py` | Dead code | Duplicate SO search at L42-45 (`origin.startswith("S")` branch does the exact same query as L38-40) |
| 4 | `mrp_production.py` | Security | Note template uses Python `str.format()` with user-controlled template string from `ir.config_parameter` -- could expose object attributes via `{0.__class__}` style injection |
| 5 | `sale_order_line.py` | OCA | Missing license header |
| 6 | `res_config_settings.py` | Convention | Reviewer `rousseldenis` already flagged: prefer `config_parameter=` attribute or a proper company-level field with `related=` instead of manual compute/inverse |

### Suggestions

| # | File | Finding |
|---|------|---------|
| 7 | `__manifest__.py` | Remove commented-out `"demo"` key |
| 8 | Tests | Excessive `invalidate_all()` + re-browse suggests misunderstanding of ORM cache -- consider simplifying |
| 9 | `mrp_production.py` | `bom_param` argument in helper is accepted but never used -- remove or implement |

rousseldenis commented:
@bosd Could you improve your commit messages and squash some of them if needed. Thanks

Commit:
[ADD] sale_order_add_byproduct

New module that automatically adds by-products produced by a Manufacturing
Order to the originating Sale Order as additional invoiceable lines.

Designed for contract manufacturing where the by-product yield is only
known after production. When an MO reaches the done state, its saleable
by-products (sale_ok=True) are added to the linked SO, with prices drawn
from the order's pricelist and an optional per-company note template.
Flagged lines (is_mrp_byproduct_line) are excluded from stock procurement
so they do not trigger additional manufacturing.

bosd commented:
@rousseldenis Done, squashed the commits.

flotho commented:
Hi @bosd ,
Any chance to rebase ?

bosd commented:
> Hi @bosd ,\nAny chance to rebase ?

Im on holiday.. will try..

/ocabot rebase

alexey-pelykh reviewed (approved):
Thanks for the thorough rework, @bosd — I went back through the current code and everything from my earlier review is addressed:

- `_action_launch_stock_rule` now excludes byproduct lines and calls `super()` on the remainder (`lines_to_procure`), so mixed recordsets procure correctly.
- The `create` override is gone — byproduct lines are kept out of procurement purely via the flag + stock-rule override.
- SO resolution goes through `sale_line_id.order_id` / `procurement_group_id.sale_id` with a single `origin` fallback (`_get_sale_order_for_byproducts`), so the duplicate search is dropped.
- The note template uses `.replace()` on known placeholders instead of `str.format()`.
- License headers are in place, and the settings field is now `related="company_id.byproduct_note_template"` on a real `res.company` field.

Nicely factored too — `_get_byproduct_moves_to_add` (with `picked` move lines + `float_compare`), the batched `create()`, and `_logger.debug`. Test coverage is solid.

Non-blocking: codecov is still red on the patch — worth a glance, but the functional suite passes and I don't treat it as a blocker.

LGTM.

Co-Reviewed-By: Claude Opus 4.8 <noreply@anthropic.com>
